src/basic/chase.c

```diff
@@ -550,19 +550,20 @@ int chase(const char *path, const char *root, ChaseFlags flags, char **ret_path,
                 return r;
 
         if (ret_path) {
-                if (!FLAGS_SET(flags, CHASE_EXTRACT_FILENAME)) {
-                        _cleanup_free_ char *q = NULL;
+                if (!FLAGS_SET(flags, CHASE_EXTRACT_FILENAME) && !empty_or_root(root)) {
+                        char *q;
 
-                        q = path_join(root, p);
+                        /* When "root" points to the root directory, the result of chaseat() is always
+                         * absolute, hence it is not necessary to prefix with the root. When "root" points to
+                         * a non-root directory, the result path is always normalized and relative, hence
+                         * we can simply call path_join() and not necessary to call path_simplify().
+                         * Note that the result of chaseat() may start with "." (more specifically, it may be
+                         * "." or "./"), and we need to drop "." in that case. */
```

Comment:
Shouldn't we change this so that we don't prefix with root if `chaseat()` returns an absolute path? So that if `CHASE_AT_RESOLVE_IN_ROOT` is not set and we follow an absolute symlink, we don't try to prefix that symlink with the root?

Reply:
Sorry, I cannot follow.

> Shouldn't we change this so that we don't prefix with root if `chaseat()` returns an absolute path?

With this change, we do not prefix with root when we get absolute path from `chaseat()`.

> So that if `CHASE_AT_RESOLVE_IN_ROOT` is not set and we follow an absolute symlink, we don't try to prefix that symlink with the root?

`chase()` unconditionally (superficially not, but effectively so) set `CHASE_AT_RESOLVE_IN_ROOT`. Note, `CHASE_AT_RESOLVE_IN_ROOT` is meaning less if `root` is `"/"`.
Hence, `chaseat()` returns relative path when `root` is not `"/"`.

Reply:
Added assertions about the returned path from `chaseat()`. PTAL.

Reply:
Let's say we dir_fd points to /abc/def and we resolve dir_fd, "qed" where qed is an absolute symlink to /usr. Now chaseat() will return an absolute path (see your earlier commit) and `root` will not be `NULL` or "/" in chase(), so an assertion will be triggered.

Reply:
In that case, `chaseat()` returns an absolute path only when `CHASE_AT_RESOLVE_IN_ROOT` is _not_ set.

If `CHASE_AT_RESOLVE_IN_ROOT` is set, the result will be `"usr"`, instead of `"/usr"`.